### ISSUE TITLE
Code editor switched to non-monotyped font after crash / No setting to select a monotyped font to fix.

### ISSUE THREAD

**Issue by 8bitprodigy** [Genio-The-Haiku-IDE/Genio#316]
Not sure what happened, but Genio locked up, and when I went to start it back up, I noticed tabs weren't indenting as far as I was used to -- turns out that somehow the font changed away from a monotyped one. I tried going through the settings, but there was no option to change the font.

**Comment by nexus6-haiku** [comment]
Could check if in Appearance (Haiku preferences) you have a monospaced font in Fixed Font?

**Comment by nexus6-haiku** [comment]
As a side note, next time Genio crashes please save a crash report and attach it to a ticket. Thank you!

**Comment by jackburton79** [comment]
> Could check if in Appearance (Haiku preferences) you have a monospaced font under Fixed Font?

I think there is no way not to have a monospaced font as a fixed font

**Comment by jackburton79** [comment]
> Not sure what happened, but Genio locked up, and when I went to start it back up, I noticed tabs weren't indenting as far as I was used to -- turns out that somehow the font changed away from a monotyped one. I tried going through the settings, but there was no option to change the font.

As far as I remember Genio uses the system fixed font, that's why we don't have a font selector preference (yet?)

**Comment by nexus6-haiku** [comment]
> > Could check if in Appearance (Haiku preferences) you have a monospaced font under Fixed Font?
> 
> I think there is no way not to have a monospaced font as a fixed font

That's correct but if you select a font which is not suitable for displaying characters or is invalid, it causes Genio to adopt a non monospaced font as a fallback, I think.
If I select Noto Color Emoji, for example, Genio seems to adopt non-monospaced font for the Editor:
![2024-02-11_13-31](https://github.com/Genio-The-Haiku-IDE/Genio/assets/105149392/eee58d9f-29b4-4e25-8ae5-0bf9e05f13bd)

**Comment by jackburton79** [comment]
I see. Don't we just use be_fixed_font, though? If so, I think this is an Haiku app_server issue, as that's the one responsible for font fallback

**Comment by jackburton79** [comment]
> I see. Don't we just use be_fixed_font, though? If so, I think this is an Haiku app_server issue, as that's the one responsible for font fallback

The font fallback should work, though, at least according to this : https://dev.haiku-os.org/ticket/15764

**Comment by nexus6-haiku** [comment]
I confirm the very same issue in Koder while using Noto Color Emoji, as well.
Let's wait for @8bitprodigy to confirm which font they have under the **Fixed font** category but this definitely sounds like a bug in Haiku and potentially the previous crash was caused by the font change, maybe.
@jackburton79 I would propose to add the option to choose a custom font in the future.

BTW, I strongly recommend using a dedicated fixed font like JetBrains Mono Nerd available on Haiku Depot.

**Comment by jackburton79** [comment]
> @jackburton79 I would propose to add the option to choose a custom font in the future.

Yeah, I agree.

**Comment by lonemadmax** [comment]
> 
> The font fallback should work, though, at least according to this : https://dev.haiku-os.org/ticket/15764

And the problem is it does, for some meaning of work that is very inadequate in this case. When the chosen font does not have a glyph the simple fallback system looks for one a hardcoded list that has. There's just enough code to keep the style (bold, italic, regular and the likes) if the checked families do have the same one as the original font, but that's it. Nothing fancy like metrics or aspect similarity, not even monospaced vs proportional. So you choose Noto Color Emoji (which you probably shouldn't [for other reasons](https://dev.haiku-os.org/ticket/16938)) that does not have letters and the system gives you Noto Sans Display Regular.

**Comment by jackburton79** [comment]
Oh I see now. That means it's really a Haiku issue. What shall we do with this issue, then?

**Comment by nexus6-haiku** [comment]
> which you probably shouldn't for other reasons

Indeed. That was just to trigger an edge case for test purpose only.
We still don't know which font is set in @8bitprodigy's case, but we can reasonably assume it's not Genio's fault.

**Comment by 8bitprodigy** [comment]
> Could check if in Appearance (Haiku preferences) you have a monospaced font under Fixed Font?

It's set to Noto Sans Mono.

> As a side note, next time Genio crashes please save a crash report and attach it to a ticket. Thank you!

Will do!

**Comment by nexus6-haiku** [comment]
> It's set to Noto Sans Mono.

If you have other fixed fonts installed try to switch to another one and restart Genio

**Comment by 8bitprodigy** [comment]
Well, I've restarted the VM since then, and that fixed the problem(for now), but I'm sure it's something that may crop up again, eventually.

**Comment by nexus6-haiku** [comment]
I reasonably believe it's not an issue with Genio but a, hopefully temporary, problem with your App Server running in a VM.
I close this for now.
